File: ruby/knight_tour.rb
def is_valid(x, y)
    if x < 0 || y < 0 || x >= N || y >= N 
        return false
    end

    return true
end

def knight_tour(visited, x, y, steps) 
    visited[x][y] = steps
    if steps == N * N 
        i = 0
        while i < N 
            j = 0
            while j < N
                printf("%d ", visited[i][j])
                j += 1
            end
            puts ""
            i += 1
        end
        puts ""
        visited[x][y] = 0
        return
    end

    i = 0
    while i < ROW.length
        new_x = x + ROW[i]
        new_y = y + COL[i]
        if is_valid(new_x,new_y) and visited[new_x][new_y] == 0 
            knight_tour(visited, new_x, new_y, steps + 1)
        end
        i += 1
    end
    visited[x][y] = 0
end

def print_visited(visited)
    i = 0
    while i < N 
        j = 0
        while j < N
            printf("%d ", visited[i][j])
            j += 1
        end
        puts ""
        i += 1
    end
    puts ""
end

N = 5
ROW = [2, 1, -1, -2, -2, -1,  1,  2]
COL = [1, 2,  2,  1, -1, -2, -2, -1]
visited = Array.new(N){Array.new(N)}
i = 0
while i < N 
    j = 0 
    while j < N 
        visited[i][j] = 0
        j += 1
    end
    i += 1
end

knight_tour(visited, 0, 0, 1)

File: cpp/knight_tour.cpp
#include <iostream>
using namespace std;

const int N = 5;
const int ROW[8] = {2, 1, -1, -2, -2, -1, 1, 2};
const int COL[8] = {1, 2, 2, 1, -1, -2, -2, -1};

bool isValid(int x, int y) {
    if(x < 0 || y < 0 || x >= N || y >= N) {
        return false;
    }

    return true;
}

void printVisited(int visited[N][N]) {
    for(int i = 0; i < N; i++) {
        for(int j = 0; j < N; j++) {
            cout<<visited[i][j]<<" ";
        }
        cout<<"\n";
    }
    cout<<"\n";
}


void knightTour(int visited[N][N], int x, int y, int steps) {
    visited[x][y] = steps;
    if(steps >= N * N) {
        printVisited(visited);
        visited[x][y] = 0;
        return;
    }

    for(int i = 0; i < 8; i++) {
        int newX = x + ROW[i];
        int newY = y + COL[i];
        if(isValid(newX, newY) && visited[newX][newY] == 0) {
            knightTour(visited, newX, newY, steps + 1);
        }
    }
    visited[x][y] = 0;
}

int main() {
    int visited[N][N];
    for(int i = 0; i < N; i++) {
        for(int j = 0; j < N; j++) {
            visited[i][j] = 0;
        }
    }
    knightTour(visited, 0, 0, 1);
    return 0;
}